File: apps/app/lib/services/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'auth_service.dart';

class ApiClient {
  static const _baseUrl = String.fromEnvironment(
    'API_URL',
    defaultValue: 'http://localhost:3001/api',
  );

  final AuthService authService;

  ApiClient(this.authService);

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        if (authService.accessToken != null) 'Authorization': 'Bearer ${authService.accessToken}',
      };

  Future<dynamic> get(String path) async {
    final response = await http.get(Uri.parse('$_baseUrl$path'), headers: _headers);
    return _decode(response);
  }

  Future<dynamic> post(String path, Map<String, dynamic> body) async {
    final response = await http.post(Uri.parse('$_baseUrl$path'), headers: _headers, body: jsonEncode(body));
    return _decode(response);
  }

  Future<dynamic> patch(String path, Map<String, dynamic> body) async {
    final response = await http.patch(Uri.parse('$_baseUrl$path'), headers: _headers, body: jsonEncode(body));
    return _decode(response);
  }

  Future<void> delete(String path) async {
    await http.delete(Uri.parse('$_baseUrl$path'), headers: _headers);
  }

  dynamic _decode(http.Response response) {
    if (response.statusCode >= 400) {
      throw Exception('Erreur API (${response.statusCode}) : ${response.body}');
    }
    if (response.body.isEmpty) return null;
    return jsonDecode(response.body);
  }
}

File: apps/web-family/lib/api-client.ts
'use client';

import { useAuth } from './auth';

const API_URL = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:3001/api';

export function useApiClient() {
  const { token } = useAuth();

  async function request<T>(path: string, init?: RequestInit): Promise<T> {
    const response = await fetch(`${API_URL}${path}`, {
      ...init,
      headers: {
        'Content-Type': 'application/json',
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
        ...init?.headers,
      },
    });
    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: 'Erreur API' }));
      throw new Error(error.message ?? 'Erreur API');
    }
    if (response.status === 204) return undefined as T;
    return response.json();
  }

  return {
    get: <T>(path: string) => request<T>(path),
    post: <T>(path: string, body?: unknown) => request<T>(path, { method: 'POST', body: JSON.stringify(body) }),
  };
}
